File: src/main/java/org/ariane/thirdlab/controller/PostController.java
package org.ariane.thirdlab.controller;

import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.util.ArrayList;
import java.util.List;

import org.ariane.thirdlab.constvalue.RtCode;
import org.ariane.thirdlab.controller.req.PostDetailReq;
import org.ariane.thirdlab.controller.resp.PostResp;
import org.ariane.thirdlab.controller.resp.PostsOfSpecTagResp;
import org.ariane.thirdlab.domain.Post;
import org.ariane.thirdlab.resp.LabResp;
import org.ariane.thirdlab.service.PostService;
import org.ariane.thirdlab.service.data.PostDetailData;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseBody;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.multipart.MultipartFile;

@RestController
@RequestMapping(value = "/api")
public class PostController {

	@Autowired
	private PostService postService;

	/**
	 * 获取所有文章
	 *
	 * @return
	 */
	@RequestMapping(value = "/posts", method = RequestMethod.GET, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<List<PostResp>> allPost() {
		List<Post> posts = postService.findAllPosts();

		// 建立临时表
		List<PostResp> pDatas = new ArrayList<>();
		for (Post post : posts) {
			PostResp postResp = new PostResp();
			postResp.id = post.getId();
			postResp.title = post.getTitle();
			postResp.parantId = post.getParantId();
			postResp.status = post.getStatus();

			pDatas.add(postResp);
		}

		LabResp<List<PostResp>> resp = new LabResp<>(RtCode.SUCCESS);
		resp.data = pDatas;
		return resp;
	}

	/**
	 * 获取所有文章
	 *
	 * @return
	 */
	@RequestMapping(value = "/tags/{tagId}/posts", method = RequestMethod.GET, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<PostsOfSpecTagResp> findPostsOfSpecTag(@PathVariable(value = "tagId") long tagId) {
		List<Post> posts = postService.findPostsByTagId(tagId);

		// 建立临时表
		List<PostResp> pDatas = new ArrayList<>();
		for (Post post : posts) {
			PostResp postResp = new PostResp();
			postResp.id = post.getId();
			postResp.title = post.getTitle();

			postResp.parantId = post.getParantId();
			postResp.status = post.getStatus();

			pDatas.add(postResp);
		}

		LabResp<PostsOfSpecTagResp> resp = new LabResp<>(RtCode.SUCCESS);
		PostsOfSpecTagResp postsOfSpecTagResp = new PostsOfSpecTagResp();
		postsOfSpecTagResp.posts = pDatas;
		postsOfSpecTagResp.tagId = tagId;
		resp.data = postsOfSpecTagResp;
		return resp;
	}

	@RequestMapping(value = "/whichpost", method = RequestMethod.GET, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<PostPositionData> searchPost(@RequestParam String postParam) {
		List<Post> posts = postService.findSpecPost(postParam);
		if (posts.size() > 0) {
			PostPositionData postPositionData = new PostPositionData();
			for (Post post : posts) {
				postPositionData.postInfos.add(new PostInfo(post.getId(), post.getTitle()));
			}

			LabResp<PostPositionData> resp = new LabResp<>(RtCode.SUCCESS);
			resp.data = postPositionData;
			return resp;

		} else {
			PostPositionData postPositionData = new PostPositionData();
			LabResp<PostPositionData> resp = new LabResp<>(RtCode.SUCCESS);
			resp.data = postPositionData;
			return resp;
		}
	}

	/**
	 * 获取某一篇文章
	 *
	 * @param id
	 * @return
	 */
	@RequestMapping(value = "/posts/{id}", method = RequestMethod.GET, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<PostDetailData> specPost(@PathVariable long id) {
		PostDetailData postDetailData = postService.findSpecPost(id);
		if (postDetailData == null) {
			LabResp<PostDetailData> resp = new LabResp<>(RtCode.POST_NOT_FOUND);
			return resp;
		}

		LabResp<PostDetailData> resp = new LabResp<>(RtCode.SUCCESS);
		resp.data = postDetailData;
		return resp;
	}

	/**
	 * 创建文章
	 *
	 * @param needCreatePost
	 * @return
	 */
	@RequestMapping(value = "/posts", method = RequestMethod.POST, consumes = "application/json", produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<PostDetailData> createPost(@RequestBody PostDetailData needCreatePost) {
		long parantId = needCreatePost.parantId;
		String title = needCreatePost.title;
		String postData = needCreatePost.postText;
		Post post = postService.createPost(parantId, title, postData);
		PostDetailData postDetailData = new PostDetailData(post);
		LabResp<PostDetailData> resp = new LabResp<>(RtCode.SUCCESS);
		resp.data = postDetailData;
		return resp;
	}

	/**
	 * 更新文章
	 *
	 * @param id
	 * @param needUpdatePost
	 * @return
	 */
	@RequestMapping(value = "/posts/{id}", method = RequestMethod.PUT, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<PostDetailData> updatePost(@PathVariable long id, @RequestBody PostDetailReq needUpdatePost) {
		long parantId = needUpdatePost.parantId;
		String title = needUpdatePost.title;
		String postText = needUpdatePost.postText;
		List<String> tags = needUpdatePost.tags;
		Post post = postService.updatePost(id, parantId, title, postText, tags);
		PostDetailData postDetailData = new PostDetailData(post);
		LabResp<PostDetailData> resp = new LabResp<>(RtCode.SUCCESS);
		resp.data = postDetailData;
		return resp;
	}

	/**
	 * 删除文章
	 *
	 * @param id
	 */
	@RequestMapping(value = "/posts/{id}", method = RequestMethod.DELETE, produces = "application/json")
	@ResponseStatus(HttpStatus.OK)
	public LabResp<Integer> deletePost(@PathVariable long id) {
		postService.deletePost(id);

		LabResp<Integer> resp = new LabResp<>(RtCode.SUCCESS);
		return resp;
	}

	@RequestMapping(value = "/picupload", method = RequestMethod.POST)
	public @ResponseBody
	String handleFileUpload(// @RequestParam("name") String
							// name,
							@RequestParam("file") MultipartFile file) {
		String name = "dddd.jpg";
		if (!file.isEmpty()) {
			try {
				byte[] bytes = file.getBytes();
				BufferedOutputStream stream = new BufferedOutputStream(new FileOutputStream(new File(name)));
				stream.write(bytes);
				stream.close();
				return "You successfully uploaded " + name + "!";
			} catch (Exception e) {
				return "You failed to upload " + name + " => " + e.getMessage();
			}
		} else {
			return "You failed to upload " + name + " because the file was empty.";
		}
	}

	// @RequestMapping("/picupload")
	// public String fileUpload2(@RequestParam("file") CommonsMultipartFile
	// file) throws IOException {
	// long startTime = System.currentTimeMillis();
	// System.out.println("fileName：" + file.getOriginalFilename());
	// String path = "./" + new Date().getTime() + file.getOriginalFilename();
	//
	// File newFile = new File(path);
	// // 通过CommonsMultipartFile的方法直接写文件（注意这个时候）
	// file.transferTo(newFile);
	// long endTime = System.currentTimeMillis();
	// System.out.println("方法二的运行时间：" + String.valueOf(endTime - startTime) +
	// "ms");
	// return "1";
	// }

	public static class PostPositionData {
		public List<PostInfo> postInfos = new ArrayList<>();
	}

	public static class PostInfo {
		public long id;
		public String title;

		public PostInfo() {

		}

		public PostInfo(long id, String title) {
			this.id = id;
			this.title = title;
		}
	}

}

File: src/main/kotlin/com/ariane/thirdlab/controller/PostController.kt
package com.ariane.thirdlab.controller

import com.ariane.thirdlab.controller.resp.TlBaseResp
import com.ariane.thirdlab.domains.Post
import com.ariane.thirdlab.repositories.PostRepository
import com.ariane.thirdlab.repositories.PostTagRepository
import com.ariane.thirdlab.rtcode.POST_NOT_FOUND
import com.ariane.thirdlab.rtcode.POST_TAG_NOT_FOUND
import com.ariane.thirdlab.rtcode.SUCCESS
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import java.util.*
import javax.transaction.Transactional

val PREPARE: Int = 1 // 预备

@RestController
@RequestMapping("/api", produces = ["application/json"])
@Transactional
open class PostController() {

    @Autowired
    lateinit var postRepository: PostRepository

    @Autowired
    lateinit var postTagRepository: PostTagRepository

    @GetMapping("/posts")
    open fun findAll(): TlBaseResp<List<PostResp>> {
        val posts = postRepository.findAll()
        val postList = mutableListOf<PostResp>()
        for (post in posts) {
            val postResp = PostResp(post.id, post.title, post.parentId, post.status, post.tagId)
            postList.add(postResp)
        }

        val resp = TlBaseResp<List<PostResp>>(SUCCESS)
        resp.data = postList

        return resp
    }

    @GetMapping("/whichpost")
    open fun findByPostParam(@RequestParam postParam: String): TlBaseResp<PostPositionResp> {
        val param = postParam.trim()
        val posts = postRepository.findByTitleContaining(param)

        val postInfos = mutableListOf<PostInfoResp>()
        for (post in posts) {
            val postInfo = PostInfoResp(post.id, post.title)
            postInfos += postInfo
        }
        val postPositionResp = PostPositionResp(postInfos)

        val resp = TlBaseResp<PostPositionResp>(SUCCESS)
        resp.data = postPositionResp

        return resp
    }

    @GetMapping("/posts/{id}")
    open fun findSpecPost(@PathVariable id: Long): TlBaseResp<PostDetailResp?> {
        val postVal = postRepository.findById(id)
        if (!postVal.isPresent) {
            val resp = TlBaseResp<PostDetailResp?>(POST_NOT_FOUND)
            return resp
        }

        val post = postVal.get()
        val detail = PostDetailResp(
            id = post.id,
            title = post.title,
            postText = post.post,
            parentId = post.parentId,
            status = post.status,
            tagId = post.tagId
        )

        val resp = TlBaseResp<PostDetailResp?>(SUCCESS)
        resp.data = detail
        return resp
    }

    @GetMapping("/tags/{tagId}/posts")
    open fun findPostsOfSpecTag(@PathVariable tagId: Long): TlBaseResp<SpecTagPostsResp> {
        val posts = if (tagId != 0L) {
            postRepository.findByTagId(tagId)
        } else {
            postRepository.findByNoTags(0)
        }

        val postListResp = mutableListOf<TagPostResp>()
        for (post in posts) {
            val tagPostResp = TagPostResp(
                id = post.id,
                title = post.title,
                parentId = post.parentId,
                status = post.status
            )
            postListResp += tagPostResp
        }
        val specTagPostsResp = SpecTagPostsResp(
            tagId = tagId,
            posts = postListResp
        )

        val resp = TlBaseResp<SpecTagPostsResp>(SUCCESS)
        resp.data = specTagPostsResp
        return resp
    }

    @PostMapping("/posts")
    open fun addPost(@RequestBody post: AddPostReq): TlBaseResp<Int?> {
        val newPost = Post()
        newPost.parentId = post.parentId
        newPost.title = post.title
        newPost.post = post.postText
        newPost.createTime = Date()
        newPost.lastModifiedTime = Date()
        newPost.status = PREPARE
        newPost.noTags = 0
        newPost.tagId = 0

        postRepository.save(newPost)

        val resp = TlBaseResp<Int?>(SUCCESS)
        resp.data = 0

        return resp
    }

    @PostMapping("/posts/{id}")
    open fun updatePost(@PathVariable id: Long, @RequestBody req: UpdatePostReq): TlBaseResp<UpdatePostResp?> {
        val postVal = postRepository.findById(id)
        if (!postVal.isPresent) {
            return TlBaseResp<UpdatePostResp?>(POST_NOT_FOUND)
        }

        val post = postVal.get()
        post.parentId = req.parentId
        post.title = req.title
        post.post = req.postText
        post.lastModifiedTime = Date()

        val resp = TlBaseResp<UpdatePostResp?>(SUCCESS)
        resp.data = UpdatePostResp(post.id)
        return resp
    }

    @PutMapping("/posts/{id}/tags/{tagId}")
    open fun movePost2NewTag(@PathVariable id: Long, @PathVariable tagId: Long): TlBaseResp<Move2NewTagResp?> {
        if (tagId != 0L) {
            val postTag = postTagRepository.findById(tagId)
            if (!postTag.isPresent) {
                return TlBaseResp<Move2NewTagResp?>(POST_TAG_NOT_FOUND)
            }
        }

        val postVal = postRepository.findById(id)
        if (!postVal.isPresent) {
            return TlBaseResp<Move2NewTagResp?>(POST_NOT_FOUND)
        }

        val post = postVal.get()
        post.tagId = tagId

        val subPosts = mutableListOf<Post>()
        searchPostAndSubPosts(subPosts, post.id)
        for (eachPost in subPosts) {
            eachPost.tagId = tagId
        }

        val resp = TlBaseResp<Move2NewTagResp?>(SUCCESS)
        resp.data = Move2NewTagResp(
            id = post.id
        )
        return resp
    }

    private fun searchPostAndSubPosts(posts: MutableList<Post>, postId: Long) {
        val subPosts = postRepository.findByParentId(postId)
        posts.addAll(subPosts)
        for (post in subPosts) {
            searchPostAndSubPosts(posts, post.id)
        }
    }

    @DeleteMapping("/posts/{id}")
    open fun delPost(@PathVariable id: Long): TlBaseResp<Int?> {
        val postVal = postRepository.findById(id)
        if (!postVal.isPresent) {
            return TlBaseResp<Int?>(POST_NOT_FOUND)
        }

        val post = postVal.get()
        postRepository.delete(post)

        val resp = TlBaseResp<Int?>(SUCCESS)
        return resp
    }
}

data class PostResp(
    val id: Long,
    val title: String,
    val parentId: Long,
    val status: Int,
    val tagId: Long
)

data class PostInfoResp(val id: Long, val title: String)
data class PostPositionResp(val postInfos: List<PostInfoResp>)

data class PostDetailResp(
    val id: Long,
    val title: String,
    val postText: String,
    val parentId: Long,
    val status: Int,
    val tagId: Long
)

data class SpecTagPostsResp(val tagId: Long, val posts: List<TagPostResp>)
data class TagPostResp(val id: Long, val title: String, val parentId: Long, val status: Int)

data class AddPostReq(val parentId: Long, val title: String, val postText: String)

data class UpdatePostReq(val parentId: Long, val title: String, val postText: String)
data class UpdatePostResp(val id: Long)

data class Move2NewTagResp(val id: Long)